libsplinter/src/network/peer_manager/mod.rs

```diff
@@ -0,0 +1,1040 @@
+            if sender.send(Ok(peers.peer_ids())).is_err() {
+                warn!("connector dropped before receiving result of list peers");
+            }
+        }
+    };
+}
+
+fn add_peer(
+    peer_id: String,
+    endpoints: Vec<String>,
+    connector: Connector,
+    peers: &mut PeerMap,
+    peer_remover: &PeerRemover,
+    ref_map: &mut RefMap,
+) -> Result<PeerRef, PeerRefAddError> {
+    let new_ref_count = ref_map.add_ref(peer_id.to_string());
+
+    // if this is not a new peer, return sucess
+    if new_ref_count > 1 {
+        let peer_ref = PeerRef::new(peer_id, peer_remover.clone());
+        return Ok(peer_ref);
+    };
+
+    debug!("Attempting to peer with {}", peer_id);
+    // If new, try to create a connection
+    for endpoint in endpoints.iter() {
+        match connector.request_connection(&endpoint) {
+            Ok(()) => {
+                debug!("Peered with {}: {}", peer_id, endpoint);
+                peers.insert(peer_id.clone(), endpoints.clone(), endpoint.to_string());
+                let peer_ref = PeerRef::new(peer_id, peer_remover.clone());
+                return Ok(peer_ref);
+            }
+            Err(err) => {
+                debug!("Unable to connect to endpoint {}: {}", endpoint, err);
+                continue;
+            }
+        }
+    }
+
+    warn!("Unable to peer with {}", peer_id);
+    // remove the reference created above
+    ref_map.remove_ref(&peer_id);
+    // unable to connect to any of the endpoints provided
+    Err(PeerRefAddError::AddError(format!(
+        "Unable to connect any endpoint that was provided for peer {}",
+        peer_id
+    )))
+}
+
+fn update_peer(
+    peer_id: String,
+    new_peer_id: String,
+    peers: &mut PeerMap,
+    ref_map: &mut RefMap,
+) -> Result<(), PeerRefUpdateError> {
+    // update the ref_map, so old PeerRef can still be used to drop references
+    if ref_map
+        .update_ref(peer_id.clone(), new_peer_id.clone())
+        .is_err()
+    {
+        return Err(PeerRefUpdateError::UpdateError(format!(
+            "Unable to update peer, {} does not exist",
+            peer_id
+        )));
+    }
+
+    // update the peer in the peer map
+    match peers.update_peer_id(peer_id.clone(), new_peer_id.clone()) {
+        Ok(()) => {
+            debug!("Updated peer id from {} to {}", peer_id, new_peer_id);
+            Ok(())
+        }
+        Err(_) => Err(PeerRefUpdateError::UpdateError(format!(
+            "Unable to update peer, {} does not exist",
+            peer_id
+        ))),
+    }
+}
+
+fn remove_peer(
+    peer_id: String,
+    connector: Connector,
+    peers: &mut PeerMap,
+    ref_map: &mut RefMap,
+) -> Result<(), PeerRefRemoveError> {
+    // remove the reference
+    let removed_peer = ref_map.remove_ref(&peer_id);
+
+    if let Some(removed_peer) = removed_peer {
+        let active_endpoint = peers.remove(&removed_peer).unwrap_or_else(|| {
+            panic!(
```

Comment:
Should we panic here?

Reply:
@vaporos  Had mentioned panicking if the ref counts are inaccurate. We probably don't need to panic here as well because the refmap will panic if we try to remove an already removed peer.

Reply:
removed